File: src/com/irisine/jiris/compiler/IrisCompiler.java
package com.irisine.jiris.compiler;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Queue;

import org.irislang.jiris.compiler.IrisInterpreter;
import org.irislang.jiris.compiler.IrisNativeJavaClass;
import org.irislang.jiris.compiler.IrisRunnable;
import org.irislang.jiris.core.IrisContextEnvironment;
import org.irislang.jiris.core.IrisThreadInfo;
import org.irislang.jiris.core.IrisValue;
import org.irislang.jiris.dev.IrisDevUtil;

import com.irisine.jiris.compiler.assistpart.IrisBlock;
import com.irisine.jiris.compiler.assistpart.IrisDeferredBlock;
import com.irisine.jiris.compiler.parser.IrisParser;
import com.irisine.jiris.compiler.parser.ParseException;
import com.irisine.jiris.compiler.statement.IrisStatement;

import net.bytebuddy.ByteBuddy;
import net.bytebuddy.asm.AsmVisitorWrapper;
import net.bytebuddy.description.method.MethodDescription;
import net.bytebuddy.description.modifier.*;
import net.bytebuddy.description.type.TypeDescription;
import net.bytebuddy.dynamic.DynamicType.Builder;
import net.bytebuddy.dynamic.DynamicType.Builder.MethodDefinition.ReceiverTypeDefinition;
import net.bytebuddy.dynamic.DynamicType.Unloaded;
import net.bytebuddy.dynamic.scaffold.InstrumentedType;
import net.bytebuddy.implementation.Implementation;
import net.bytebuddy.implementation.Implementation.Context;
import net.bytebuddy.implementation.bytecode.ByteCodeAppender;
import net.bytebuddy.jar.asm.ClassVisitor;
import net.bytebuddy.jar.asm.ClassWriter;
import net.bytebuddy.jar.asm.Label;
import net.bytebuddy.jar.asm.MethodVisitor;
import net.bytebuddy.jar.asm.Opcodes;

public enum IrisCompiler {
	
	INSTANCE;
	
	public enum CurrentDefineType {
		Normal,
		Class,
		Module,
		Interface,
	}
	
	private String m_currentFile = "";
	private ByteBuddy m_curretByteBuddy = null; 
	private String m_currentClassName = "";
		
	private LinkedList<IrisStatement> m_statements = new LinkedList<IrisStatement>();
	private Builder<IrisNativeJavaClass> m_currentBuilder = null;
	
	private LinkedList<String> m_uniqueStrings = new LinkedList<String>();
	private HashMap<String, Integer> m_uniqueStringHash = new HashMap<String, Integer>();
	
	private Queue<IrisDeferredBlock> m_deferredStatements = new LinkedList<IrisDeferredBlock>();
	private HashMap<String, Integer> m_blockNameCount = new HashMap<String ,Integer>();
	
	private Class<?> m_javaClass = null;
	
	private CurrentDefineType m_currentDefineType = CurrentDefineType.Normal;
	private String m_currentDefineName = "";
	
	private boolean m_staticDefine = false;
	
	private Label m_currentEndLable = null;
	
	private boolean m_firstStackFrameGenerated = false;
	
	public boolean isFirstStackFrameGenerated() {
		return m_firstStackFrameGenerated;
	}
	
	public void setFirstStackFrameGenerated(boolean firstStackFrameGenerated) {
		m_firstStackFrameGenerated = firstStackFrameGenerated;
	}
	
	public Label getCurrentEndLable() {
		return m_currentEndLable;
	}
	
	private void setCurrentEndLable(Label currentEndLable) {
		m_currentEndLable = currentEndLable;
	}
	
	public void setStaticDefine(boolean staticDefine) {
		m_staticDefine = staticDefine;
	}
	
	public int GetIndexOfContextVar() {
		return m_staticDefine ? 0 : 1;
	}
	
	public int GetIndexOfThreadInfoVar() {
		return GetIndexOfContextVar() + 1;
	}
	
	public int GetIndexOfResultValue() {
		return GetIndexOfContextVar() + 2;
	}
	
	public CurrentDefineType getCurrentDefineType() {
		return m_currentDefineType;
	}
	
	public void setCurrentDefineType(CurrentDefineType currentDefineType) {
		m_currentDefineType = currentDefineType;
	}
	
	public String getCurrentDefineName() {
		return m_currentDefineName;
	}
	
	public void setCurrentDefineName(String currentDefineName) {
		m_currentDefineName = currentDefineName;
	}
	
	public int GetBlockNameCount(String name) {
		Integer count = m_blockNameCount.get(name);
		if(count == null) {
			m_blockNameCount.put(name, 1);
			return 1;
		}
		else {
			m_blockNameCount.put(name, count + 1);
			return count;
		}
	}
	
	public void PushDeferredStatement(IrisDeferredBlock statement) {
		m_deferredStatements.add(statement);
	}
	
	public IrisDeferredBlock PopDeferredStatement() {
		return m_deferredStatements.poll();
	}
	
	public boolean IsDeferredStatementQueueEmpty() {
		return m_deferredStatements.isEmpty();
	}
		
	public void AddUniqueString(String str) {
		if(!m_uniqueStringHash.containsKey(str)) {
			m_uniqueStrings.add(str);
			m_uniqueStringHash.put(str, m_uniqueStrings.size() - 1);
		}
	}
	
	public int GetUinqueIndex(String str){ 
		return m_uniqueStringHash.get(str);
	}
	
	public void AddStatement(IrisStatement statement) {
		m_statements.add(statement);
	}
	
	public LinkedList<IrisStatement> getStatements() {
		return m_statements;
	}
	
	public Builder<IrisNativeJavaClass> getCurrentBuilder() {
		return m_currentBuilder;
	}
	
	public String getCurrentClassName() {
		return m_currentClassName;
	}
	
	private boolean Generate() {
		
		int index = m_currentFile.lastIndexOf(".");
		IrisInterpreter.INSTANCE.InceamJavaClassFileNumber();
		m_currentClassName = m_currentFile.substring(0, index) + "$" + "ir";
		
		m_curretByteBuddy = new ByteBuddy();
		m_currentBuilder = m_curretByteBuddy
				.subclass(IrisNativeJavaClass.class)
				.implement(IrisRunnable.class)
				.visit(new AsmVisitorWrapper(){
			@Override
			public int mergeReader(int arg0) {
				return 0;
			}

			@Override
			public int mergeWriter(int arg0) {
				return ClassWriter.COMPUTE_FRAMES;
			}

			@Override
			public ClassVisitor wrap(TypeDescription arg0, ClassVisitor arg1, int arg2, int arg3) {
				return arg1;
			}
		}).name(m_currentClassName);
		
		m_currentBuilder = m_currentBuilder.defineField("sm_uniqueStringObjects", 
				ArrayList.class, 
				Visibility.PUBLIC, 
				FieldManifestation.FINAL, 
				Ownership.STATIC).initializer(new ByteCodeAppender(){

			@Override
			public Size apply(MethodVisitor mv, Context arg1, MethodDescription arg2) {
				mv.visitTypeInsn(Opcodes.NEW, "java/util/ArrayList");
				mv.visitInsn(Opcodes.DUP);
				mv.visitMethodInsn(Opcodes.INVOKESPECIAL, "java/util/ArrayList", "<init>", "()V", false);
				mv.visitFieldInsn(Opcodes.PUTSTATIC, m_currentClassName, "sm_uniqueStringObjects", "Ljava/util/ArrayList;");
				mv.visitInsn(Opcodes.RETURN);
				return new Size(0, 0);
			}
		});
		
		ReceiverTypeDefinition<IrisNativeJavaClass> mainDefininition = GenerateIrisNativeJavaMethod("run", new Implementation() {
			@Override
			public InstrumentedType prepare(InstrumentedType arg0) {
				return arg0;
			}

			@Override
			public ByteCodeAppender appender(Target arg0) {
				return RunMethodAppender();
			}
		});
		
		while(!IsDeferredStatementQueueEmpty()) {
			IrisDeferredBlock deferredStatement = PopDeferredStatement();
			
			mainDefininition = mainDefininition.defineMethod(deferredStatement.getGenerateName(),
					IrisValue.class, 
					Visibility.PUBLIC, 
					Ownership.STATIC)
					.withParameters(IrisContextEnvironment.class, IrisThreadInfo.class)
					.throwing(Throwable.class)
					.intercept(new Implementation() {
						@Override
						public InstrumentedType prepare(InstrumentedType arg0) {
							return arg0;
						}

						@Override
						public ByteCodeAppender appender(Target arg0) {
							return UserMethodAppender(deferredStatement.getStatement());
						}
					});
		}
		
		Unloaded<IrisNativeJavaClass> classObj = mainDefininition.make();
		
		m_javaClass = classObj.load(getClass().getClassLoader()).getLoaded();
		InitUniqueStringList(m_javaClass);
		
		byte[] jvmcode = classObj.getBytes();
		
		File file = new File(m_currentClassName + ".class");
		if(file.exists()) {
			file.delete();
		}
		FileOutputStream fos;
		try {
			fos = new FileOutputStream(file);
			fos.write(jvmcode);
			fos.close();
		} catch (FileNotFoundException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		
		return true;
	}
	
	@SuppressWarnings("unchecked")
	private void InitUniqueStringList(Class<?> javaClass) {
		ArrayList<IrisValue> values = null;
		try {
			values = (ArrayList<IrisValue>) javaClass.getField("sm_uniqueStringObjects").get(null);
		} catch (IllegalArgumentException | IllegalAccessException | NoSuchFieldException | SecurityException e1) {
			// TODO Auto-generated catch block
			e1.printStackTrace();
		}
		
		for(String uString : m_uniqueStrings) {
			values.add(IrisDevUtil.CreateUniqueString(uString));
		}
	}
	
	private ByteCodeAppender RunMethodAppender() {
		return new RunMethodAppender(this);
	}
	
	private ByteCodeAppender UserMethodAppender(IrisBlock block) {
		return new UserMethodAppender(this, block);
	}
	
	private ReceiverTypeDefinition<IrisNativeJavaClass> GenerateIrisNativeJavaMethod(String methodName, Implementation impl) {
		
		return m_currentBuilder.defineMethod(methodName, IrisValue.class, Visibility.PUBLIC)
						.withParameters(IrisContextEnvironment.class, IrisThreadInfo.class)
						.throwing(Throwable.class)
						.intercept(impl);
	}
	
	public boolean LoadScriptFromPath(String path) {
		m_currentFile = path;
		IrisParser parser = new IrisParser(path);
		try {
			parser.translation_unit();
		}
		catch (Exception e) {
			System.out.println(e.getMessage());
			//e.printStackTrace();
			parser.ReInit(System.in);
			return false;
		}
		catch (Error e) {
			System.out.println(e.getMessage());
			return false;
		}
		
		return Generate();
	}
	
	public boolean TestLoad(String path) {
		IrisParser parser = new IrisParser(path);
		try {
			parser.translation_unit();
		} catch (ParseException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		return true;
	}
	
	public Class<?> getJavaClass() {
		return m_javaClass;
	}
	
	private static class RunMethodAppender implements ByteCodeAppender {
		
		IrisCompiler m_compiler = null;
		
		public RunMethodAppender(IrisCompiler compiler) {
			m_compiler = compiler;
		}
		
		@Override
		public Size apply(MethodVisitor mv, Context arg1, MethodDescription arg2) {
			m_compiler.setFirstStackFrameGenerated(false);
			m_compiler.setStaticDefine(false);
			Label lableFrom = new Label();
			Label lableEnd = new Label();
			m_compiler.setCurrentEndLable(new Label());
			
			mv.visitMethodInsn(Opcodes.INVOKESTATIC, "org/irislang/jiris/dev/IrisDevUtil", "Nil", "()Lorg/irislang/jiris/core/IrisValue;", false);
			mv.visitVarInsn(Opcodes.ASTORE, 3);
			
			mv.visitLabel(lableFrom);
			for(IrisStatement statement : m_compiler.getStatements()) {
				statement.Generate(m_compiler, m_compiler.getCurrentBuilder(), mv);
			}
			
			mv.visitLabel(m_compiler.getCurrentEndLable());
			
			mv.visitVarInsn(Opcodes.ALOAD, 3);
			mv.visitInsn(Opcodes.ARETURN);
			
			mv.visitLabel(lableEnd);
			m_compiler.setCurrentEndLable(null);
			
			mv.visitLocalVariable("this", "L" + m_compiler.getCurrentClassName() + ";", null, lableFrom, lableEnd, 0);
			mv.visitLocalVariable("context", "Lorg/irislang/jiris/core/IrisContextEnvironment;", null, lableFrom, lableEnd, 1);
			mv.visitLocalVariable("threadInfo", "Lorg/irislang/jiris/core/IrisThreadInfo;", null, lableFrom, lableEnd, 2);
			mv.visitLocalVariable("resultValue", "Lorg/irislang/jiris/core/IrisValue;", null, lableFrom, lableEnd, 3);
			
			return new Size(0, 0);
		}
		
	}
	
	private static class UserMethodAppender implements ByteCodeAppender {
		
		IrisCompiler m_compiler = null;
		IrisBlock m_block = null;
		
		public UserMethodAppender(IrisCompiler compiler, IrisBlock block) {
			m_compiler = compiler;
			m_block = block;
		}

		@Override
		public Size apply(MethodVisitor mv, Context arg1, MethodDescription arg2) {
			m_compiler.setFirstStackFrameGenerated(false);
			m_compiler.setStaticDefine(true);
			Label lableFrom = new Label();
			Label lableEnd = new Label();
			m_compiler.setCurrentEndLable(new Label());
			
			mv.visitMethodInsn(Opcodes.INVOKESTATIC, "org/irislang/jiris/dev/IrisDevUtil", "Nil", "()Lorg/irislang/jiris/core/IrisValue;", false);
			mv.visitVarInsn(Opcodes.ASTORE, 2);
			
			mv.visitLabel(lableFrom);
			
			m_block.Generate(m_compiler, m_compiler.getCurrentBuilder(), mv);
			
			mv.visitLabel(m_compiler.getCurrentEndLable());
			m_compiler.setCurrentEndLable(null);
			
			mv.visitVarInsn(Opcodes.ALOAD, 2);
			mv.visitInsn(Opcodes.ARETURN);
			
			mv.visitLabel(lableEnd);
			
			mv.visitLocalVariable("context", "Lorg/irislang/jiris/core/IrisContextEnvironment;", null, lableFrom, lableEnd, 0);
			mv.visitLocalVariable("threadInfo", "Lorg/irislang/jiris/core/IrisThreadInfo;", null, lableFrom, lableEnd, 1);
			mv.visitLocalVariable("resultValue", "Lorg/irislang/jiris/core/IrisValue;", null, lableFrom, lableEnd, 2);
			
			return new Size(0, 0);
		}
		
	}
}

File: src/main/java/com/irisine/jiris/compiler/IrisCompiler.kt
package com.irisine.jiris.compiler

import java.io.File
import java.io.FileNotFoundException
import java.io.FileOutputStream
import java.io.IOException
import java.util.ArrayList
import java.util.HashMap
import java.util.LinkedList

import net.bytebuddy.description.field.FieldDescription
import net.bytebuddy.description.field.FieldList
import net.bytebuddy.description.method.MethodList
import net.bytebuddy.pool.TypePool
import org.irislang.jiris.IrisInterpreter
import org.irislang.jiris.IrisNativeJavaClass
import org.irislang.jiris.IrisRunnable
import org.irislang.jiris.core.IrisContextEnvironment
import org.irislang.jiris.core.IrisThreadInfo
import org.irislang.jiris.core.IrisValue
import org.irislang.jiris.core.exceptions.IrisExceptionBase
import org.irislang.jiris.dev.IrisDevUtil

import com.irisine.jiris.compiler.assistpart.IrisBlock
import com.irisine.jiris.compiler.assistpart.IrisDeferredBlock
import com.irisine.jiris.compiler.parser.IrisParser
import com.irisine.jiris.compiler.parser.ParseException
import com.irisine.jiris.compiler.statement.IrisStatement

import net.bytebuddy.ByteBuddy
import net.bytebuddy.asm.AsmVisitorWrapper
import net.bytebuddy.description.method.MethodDescription
import net.bytebuddy.description.modifier.*
import net.bytebuddy.description.type.TypeDescription
import net.bytebuddy.dynamic.DynamicType.Builder
import net.bytebuddy.dynamic.DynamicType.Builder.MethodDefinition.ReceiverTypeDefinition
import net.bytebuddy.dynamic.scaffold.InstrumentedType
import net.bytebuddy.implementation.Implementation
import net.bytebuddy.implementation.Implementation.Context
import net.bytebuddy.implementation.bytecode.ByteCodeAppender
import net.bytebuddy.jar.asm.ClassVisitor
import net.bytebuddy.jar.asm.ClassWriter
import net.bytebuddy.jar.asm.Label
import net.bytebuddy.jar.asm.MethodVisitor
import net.bytebuddy.jar.asm.Opcodes

enum class IrisCompiler {

    INSTANCE;

    private var m_currentFile = ""
    private var m_curretByteBuddy: ByteBuddy? = null
    var currentClassName = ""
        private set

    val statements = LinkedList<IrisStatement>()
    var currentBuilder: Builder<IrisNativeJavaClass>? = null
        private set

    private val m_uniqueStrings = LinkedList<String>()
    private val m_uniqueStringHash = HashMap<String, Int>()

    private val m_deferredStatements = LinkedList<IrisDeferredBlock>()
    private val m_blockNameCount = HashMap<String, Int>()

    var nativeJavaClass: Class<*>? = null
        private set

    var currentDefineType = CurrentDefineType.Normal
    var currentDefineName = ""

    var isStaticDefine = false

    var currentEndLable: Label? = null
        private set

    val irregularVariableLabelPairs = LinkedList<IrregularVariableLabelPair>()

    var currentLoopContinueLable: Label? = null

    var currentLoopEndLable: Label? = null

    var isFirstStackFrameGenerated = false

    enum class CurrentDefineType {
        Normal,
        Class,
        Module,
        Interface
    }

    class IrregularVariableLabelPair(val fromLabel: Label, val toLabel: Label)

    fun AddIrregularVariableLabelPair(pair: IrregularVariableLabelPair) {
        irregularVariableLabelPairs.push(pair)
    }

    fun ClearIrregularVariableLabelPair() {
        irregularVariableLabelPairs.clear()
    }

    fun GetIndexOfContextVar(): Int {
        return if (isStaticDefine) 0 else 1
    }

    fun GetIndexOfThreadInfoVar(): Int {
        return GetIndexOfContextVar() + 1
    }

    fun GetIndexOfResultValue(): Int {
        return GetIndexOfContextVar() + 2
    }

    fun GetIndexOfIrregularVar(): Int {
        return GetIndexOfContextVar() + 3
    }

    fun GetBlockNameCount(name: String): Int {
        val count = m_blockNameCount[name]
        if (count == null) {
            m_blockNameCount[name] = 1
            return 1
        } else {
            m_blockNameCount[name] = count + 1
            return count + 1
        }
    }

    fun PushDeferredStatement(statement: IrisDeferredBlock) {
        m_deferredStatements.add(statement)
    }

    fun PopDeferredStatement(): IrisDeferredBlock? {
        return m_deferredStatements.poll()
    }

    fun IsDeferredStatementQueueEmpty(): Boolean {
        return m_deferredStatements.isEmpty()
    }

    fun AddUniqueString(str: String) {
        if (!m_uniqueStringHash.containsKey(str)) {
            m_uniqueStrings.add(str)
            m_uniqueStringHash[str] = m_uniqueStrings.size - 1
        }
    }

    fun GetUinqueIndex(str: String): Int? {
        return m_uniqueStringHash[str]
    }

    fun AddStatement(statement: IrisStatement) {
        statements.add(statement)
    }

    private fun Generate(): Boolean {

        val index = m_currentFile.lastIndexOf(".")
        val index2 = m_currentFile.lastIndexOf("/")
        IrisInterpreter.INSTANCE.InceamJavaClassFileNumber()
        val filePath = m_currentFile.substring(0, index2 + 1)
        currentClassName = m_currentFile.substring(index2 + 1, index) + "$" + "ir"

        m_curretByteBuddy = ByteBuddy()
        currentBuilder = m_curretByteBuddy!!
                .subclass(IrisNativeJavaClass::class.java)
                .implement(IrisRunnable::class.java)
                .visit(object : AsmVisitorWrapper {
                    override fun mergeReader(arg0: Int): Int {
                        return 0
                    }

                    override fun wrap(instrumentedType: TypeDescription, classVisitor: ClassVisitor, implementationContext: Context, typePool: TypePool, fields: FieldList<FieldDescription.InDefinedShape>, methods: MethodList<*>, writerFlags: Int, readerFlags: Int): ClassVisitor {
                        return classVisitor
                    }

                    override fun mergeWriter(arg0: Int): Int {
                        return ClassWriter.COMPUTE_FRAMES
                    }

                    //			@Override
                    //			public ClassVisitor wrap(TypeDescription arg0, ClassVisitor arg1, int arg2, int arg3) {
                    //				return arg1;
                    //			}
                }).name(currentClassName)

        currentBuilder = currentBuilder!!
                .defineField("sm_uniqueStringObjects",
                        ArrayList::class.java,
                        Visibility.PUBLIC,
                        FieldManifestation.FINAL,
                        Ownership.STATIC)
                .defineField("sm_scriptFileName",
                        String::class.java,
                        Visibility.PUBLIC,
                        FieldManifestation.FINAL,
                        Ownership.STATIC)
                .initializer { mv, arg1, arg2 ->
                    mv.visitTypeInsn(Opcodes.NEW, "java/util/ArrayList")
                    mv.visitInsn(Opcodes.DUP)
                    mv.visitMethodInsn(Opcodes.INVOKESPECIAL, "java/util/ArrayList", "<init>", "()V", false)
                    mv.visitFieldInsn(Opcodes.PUTSTATIC, currentClassName, "sm_uniqueStringObjects", "Ljava/util/ArrayList;")

                    mv.visitLdcInsn(m_currentFile)
                    mv.visitFieldInsn(Opcodes.PUTSTATIC, currentClassName, "sm_scriptFileName",
                            "Ljava/lang/String;")

                    mv.visitInsn(Opcodes.RETURN)
                    ByteCodeAppender.Size(0, 0)
                }

        var mainDefininition = GenerateIrisNativeJavaMethod("run", object : Implementation {
            override fun prepare(arg0: InstrumentedType): InstrumentedType {
                return arg0
            }

            override fun appender(arg0: Implementation.Target): ByteCodeAppender {
                return RunMethodAppender()
            }
        })

        while (!IsDeferredStatementQueueEmpty()) {
            val deferredStatement = PopDeferredStatement()

            mainDefininition = mainDefininition.defineMethod(deferredStatement!!.generateName,
                    IrisValue::class.java,
                    Visibility.PUBLIC,
                    Ownership.STATIC)
                    .withParameters(IrisContextEnvironment::class.java, IrisThreadInfo::class.java)
                    .throwing(IrisExceptionBase::class.java)
                    .intercept(object : Implementation {
                        override fun prepare(arg0: InstrumentedType): InstrumentedType {
                            return arg0
                        }

                        override fun appender(arg0: Implementation.Target): ByteCodeAppender {
                            return UserMethodAppender(deferredStatement.block)
                        }
                    })
        }

        val classObj = mainDefininition.make()

        nativeJavaClass = classObj.load(javaClass.getClassLoader()).loaded
        InitUniqueStringList(nativeJavaClass!!)

        val jvmcode = classObj.bytes

        val file = File("$filePath$currentClassName.class")
        if (file.exists()) {
            file.delete()
        }
        val fos: FileOutputStream
        try {
            fos = FileOutputStream(file)
            fos.write(jvmcode)
            fos.close()
        } catch (e: FileNotFoundException) {
            // TODO Auto-generated catch block
            e.printStackTrace()
        } catch (e: IOException) {
            // TODO Auto-generated catch block
            e.printStackTrace()
        }

        return true
    }

    private fun InitUniqueStringList(javaClass: Class<*>) {
        var values: ArrayList<IrisValue>? = null
        try {
            values = javaClass.getField("sm_uniqueStringObjects").get(null) as ArrayList<IrisValue>
        } catch (e1: IllegalArgumentException) {
            // TODO Auto-generated catch block
            e1.printStackTrace()
        } catch (e1: IllegalAccessException) {
            e1.printStackTrace()
        } catch (e1: NoSuchFieldException) {
            e1.printStackTrace()
        } catch (e1: SecurityException) {
            e1.printStackTrace()
        }

        for (uString in m_uniqueStrings) {
            values!!.add(IrisDevUtil.CreateUniqueString(uString))
        }
    }

    private fun RunMethodAppender(): ByteCodeAppender {
        return RunMethodAppender(this)
    }

    private fun UserMethodAppender(block: IrisBlock): ByteCodeAppender {
        return UserMethodAppender(this, block)
    }

    private fun GenerateIrisNativeJavaMethod(methodName: String, impl: Implementation): ReceiverTypeDefinition<IrisNativeJavaClass> {

        return currentBuilder!!.defineMethod(methodName, IrisValue::class.java, Visibility.PUBLIC)
                .withParameters(IrisContextEnvironment::class.java, IrisThreadInfo::class.java)
                .throwing(IrisExceptionBase::class.java)
                .intercept(impl)
    }

    fun LoadScriptFromPath(path: String): Boolean {
        m_currentFile = path
        val parser = IrisParser(path)
        try {
            parser.translation_unit()
        } catch (e: Exception) {
            println(e.message)
            //e.printStackTrace();
            parser.ReInit(System.`in`)
            return false
        } catch (e: Error) {
            println(e.message)
            return false
        }

        return Generate()
    }

    fun TestLoad(path: String): Boolean {
        val parser = IrisParser(path)
        try {
            parser.translation_unit()
        } catch (e: ParseException) {
            // TODO Auto-generated catch block
            e.printStackTrace()
        }

        return true
    }

    private class RunMethodAppender(val compiler: IrisCompiler) : ByteCodeAppender {

        override fun apply(mv: MethodVisitor, arg1: Context, arg2: MethodDescription): ByteCodeAppender.Size {

            compiler.ClearIrregularVariableLabelPair()

            compiler.isFirstStackFrameGenerated = false
            compiler.isStaticDefine = false
            val lableFrom = Label()
            val lableEnd = Label()
            compiler.currentEndLable = Label()

            IrisGenerateHelper.SetFileName(mv, compiler)

            mv.visitMethodInsn(Opcodes.INVOKESTATIC, "org/irislang/jiris/dev/IrisDevUtil", "Nil",
                    "()Lorg/irislang/jiris/core/IrisValue;", false)
            mv.visitVarInsn(Opcodes.ASTORE, 3)

            mv.visitLabel(lableFrom)
            for (statement in compiler.statements) {
                statement.Generate(compiler, compiler.currentBuilder!!, mv)
            }

            mv.visitLabel(compiler.currentEndLable)

            mv.visitVarInsn(Opcodes.ALOAD, 3)
            mv.visitInsn(Opcodes.ARETURN)

            mv.visitLabel(lableEnd)
            compiler.currentEndLable = null

            mv.visitLocalVariable("this", "L" + compiler.currentClassName + ";", null, lableFrom, lableEnd, 0)
            mv.visitLocalVariable("context", "Lorg/irislang/jiris/core/IrisContextEnvironment;", null, lableFrom, lableEnd, 1)
            mv.visitLocalVariable("threadInfo", "Lorg/irislang/jiris/core/IrisThreadInfo;", null, lableFrom, lableEnd, 2)
            mv.visitLocalVariable("resultValue", "Lorg/irislang/jiris/core/IrisValue;", null, lableFrom, lableEnd, 3)

            for (pair in compiler.irregularVariableLabelPairs) {
                mv.visitLocalVariable("runtimeIrregular",
                        "Lorg/irislang/jiris/core/exceptions/IrisRuntimeException;", null, pair.fromLabel,
                        pair.toLabel, 4)
            }

            return ByteCodeAppender.Size(0, 0)
        }

    }

    private class UserMethodAppender(compiler: IrisCompiler, block: IrisBlock) : ByteCodeAppender {

        internal var m_compiler: IrisCompiler? = null
        internal var m_block: IrisBlock? = null

        init {
            m_compiler = compiler
            m_block = block
        }

        override fun apply(mv: MethodVisitor, arg1: Context, arg2: MethodDescription): ByteCodeAppender.Size {

            m_compiler!!.ClearIrregularVariableLabelPair()

            m_compiler!!.isFirstStackFrameGenerated = false
            m_compiler!!.isStaticDefine = true
            val lableFrom = Label()
            val lableEnd = Label()
            m_compiler!!.currentEndLable = Label()

            mv.visitMethodInsn(Opcodes.INVOKESTATIC, "org/irislang/jiris/dev/IrisDevUtil", "Nil",
                    "()Lorg/irislang/jiris/core/IrisValue;", false)
            mv.visitVarInsn(Opcodes.ASTORE, 2)

            mv.visitLabel(lableFrom)

            m_block!!.Generate(m_compiler!!, m_compiler!!.currentBuilder!!, mv)

            mv.visitLabel(m_compiler!!.currentEndLable)
            m_compiler!!.currentEndLable = null

            mv.visitVarInsn(Opcodes.ALOAD, 2)
            mv.visitInsn(Opcodes.ARETURN)

            mv.visitLabel(lableEnd)

            mv.visitLocalVariable("context", "Lorg/irislang/jiris/core/IrisContextEnvironment;",
                    null, lableFrom, lableEnd, 0)
            mv.visitLocalVariable("threadInfo", "Lorg/irislang/jiris/core/IrisThreadInfo;",
                    null, lableFrom, lableEnd, 1)
            mv.visitLocalVariable("resultValue", "Lorg/irislang/jiris/core/IrisValue;",
                    null, lableFrom, lableEnd, 2)

            for (pair in m_compiler!!.irregularVariableLabelPairs) {
                mv.visitLocalVariable("runtimeIrregular",
                        "Lorg/irislang/jiris/core/exceptions/IrisRuntimeException;", null, pair.fromLabel,
                        pair.toLabel, 3)
            }

            return ByteCodeAppender.Size(0, 0)
        }

    }
}
